### DISCUSSION TITLE
Does Expr have Control flow

### DISCUSSION THREAD

**Discussion by pat3icki** [antonmedv/expr#601] [General] [answered]
Hey, good day am trying to integrate this into a project as a middleware for the project, I predicted the script is gonna get a little complex, is there a way to have if statements or if-else, I can’t find any documentation or issues about this.
Thanks

**Comment by daenney** [comment]
As far as I know, Expr only has the ternary statement:

```
valueA != valueB ? do-if-true : do-if-false
```

It's in the [Operators](https://expr-lang.org/docs/language-definition#operators).

**Comment by antonmedv** [comment]
Yeap. But I do want to implement if statements as well.

**Comment by ctcx** [comment]
If it could support control flow and custom functions, that would be great.

**Reply by antonmedv** [reply]
Custom functions already supported.

**Comment by phuslu** [comment]
maybe we can implement a simple "pattern matching" first?

```
match a {
    1 -> "foo"
    2 -> "bar"
    default -> "none"
}
```

or

```
switch {
case <condition 1> -> "foo"
case <condition 2> -> "bar"
default -> "none"
}
```

something like aboves is really help

**Comment by phuslu** [comment]
BTW: currently I use `go template` + [sprig ](https://github.com/Masterminds/sprig) for implement my business logic, it's too heavy and a bit slow than `expr`. The blocker of my `expr` migration is if statement or pattern matching.

**Reply by antonmedv** [reply]
So what king of the problem are you trying to solve? Please, more context.

**Reply by phuslu** [reply]
Here an example of my side project, https://github.com/phuslu/liner/blob/master/example.yaml#L47-L66 , please the "policy" and "dialer" field.

**Reply by phuslu** [reply]
k8s helm also use same idiom, see https://helm.sh/docs/chart_template_guide/control_structures/

**Comment by pelletier197** [comment]
Would greatly benefit control flow as well. Either switch case or if/else are good options for my use-cases. I can work around with ternary however, just less practical.

**Comment by rchougule** [comment]
@antonmedv Any plans on having if-else in near future? It's easier for business folks as well to use

**Reply by antonmedv** [reply]
Yes, we are planning to implement if-else in near future. I guess this will be a new feature after current release.

**Comment by xiaoas** [comment]
I'm facing a scenario where if-else proves very helpful, consider this script:
```go
if input1 != "" {
  return input1
}
b := calculateB(input2);
c := calculateC();
return getResult(b, c);
```
To do this in go expr, I'm currently planning to write this:
```go
let b = input1 == "" ? calculateB(input2) : nil;
let c = input1 == "" ? calculateC() : nil;
input1 != "" ? input1 : getResult(b, c)
```
It would be so much nicer if the expression language has a "return" keyword, or a rust-like if-else expression:
```go
if input1 != "" {
  input1
} else {
  let b = calculateB(input2);
  let c = calculateC();
  getResult(b, c)
}
```
I would say this is easier to read.

**Reply by antonmedv** [reply]
True! This is why we are working on it. ping @OlgaNovg ;)

**Reply by pagrus7** [reply]
https://github.com/expr-lang/expr/pull/736
Looks like this pull request has been open few months ago. Is there a path forward?
(I'm evaluating expr for a use case which needs control flow)

**Reply by antonmedv** [reply]
Yes! @OlgaNovg is also waiting long time already)

**Comment by pagrus7** [comment]
Hey folks - excited that this is now merged! Is there a timeline for the next release?

**Reply by antonmedv** [reply]
Soon! I have to fix a few issues before. And i want to add support for sequence of expressions.

**Comment by antonmedv** [comment, marked as answer]
Now Expr has control flows!

```elm
let foo = if bar > 0 { 
    true
} else {
    false
}
```

**Reply by sonnysideup** [reply]
where is the documented?

**Reply by Ariyn** [reply]
you can read this in [1.17.0 release note](https://github.com/expr-lang/expr/releases/tag/v1.17.0). 

Very small and short, but there is a document too. (Maybe no one made a detailed document PR yet)
<img width="577" height="95" alt="스크린샷 2025-08-19 오전 11 55 49" src="https://github.com/user-attachments/assets/ec94ffe2-0a33-4455-a1c1-3648ae18a2b2" />
